core/src/VolatilityOracle.sol

```diff
@@ -0,0 +1,181 @@
+            if (timeSinceLastWrite < FEE_GROWTH_GLOBALS_SAMPLE_PERIOD) {
+                (uint160 sqrtMeanPriceX96, ) = Oracle.consult(pool, ORACLE_LOOKBACK);
+                return (sqrtMeanPriceX96, lastWrite.iv);
+            }
+
+            // Prepare to call all getters
+            Volatility.FeeGrowthGlobals[60] storage arr = feeGrowthGlobals[pool];
+
+            // Call all getters
+            Volatility.PoolData memory data = _getPoolData(pool);
+            Volatility.FeeGrowthGlobals memory a = _getFeeGrowthGlobalsOld(arr, lastWrite.index);
+            Volatility.FeeGrowthGlobals memory b = _getFeeGrowthGlobalsNow(pool);
+
+            // Default to using the existing IV
+            uint256 iv = lastWrite.iv;
+            // Only update IV if the feeGrowthGlobals samples are approximately 1 hour apart
+            if (
+                isInInterval({
+                    min: 1 hours - 5 minutes, // NOTE: Keeping these constants here because they're related to `arr.length`
+                    x: b.timestamp - a.timestamp,
+                    max: 1 hours + 5 minutes
+                })
+            ) {
+                // Estimate, then clamp so it lies within [previous - maxChange, previous + maxChange]
+                iv = Volatility.estimate(cachedMetadata[pool], data, a, b, IV_SCALE);
+
+                uint256 maxChange = timeSinceLastWrite * IV_CHANGE_PER_SECOND;
+                if (iv > lastWrite.iv + maxChange) iv = lastWrite.iv + maxChange;
+                else if (iv + maxChange < lastWrite.iv) iv = lastWrite.iv - maxChange;
+            }
+
+            // Store the new feeGrowthGlobals sample and update `lastWrites`
+            uint8 next = uint8((lastWrite.index + 1) % 60);
+            arr[next] = b;
+            lastWrites[pool] = LastWrite(next, uint32(block.timestamp), uint216(iv));
+
+            return (data.sqrtMeanPriceX96, iv);
+        }
+    }
+
+    function consult(IUniswapV3Pool pool) external view returns (uint160, uint256) {
+        (uint160 sqrtMeanPriceX96, ) = Oracle.consult(pool, ORACLE_LOOKBACK);
+        return (sqrtMeanPriceX96, lastWrites[pool].iv);
+    }
+
+    function _getPoolMetadata(IUniswapV3Pool pool) private view returns (Volatility.PoolMetadata memory metadata) {
+        (, , uint16 observationIndex, uint16 observationCardinality, , uint8 feeProtocol, ) = pool.slot0();
+        // If block times are inconsistent, `maxSecondsAgo` from oracle may be inflated. Divide by 2 to be extra safe.
+        metadata.maxSecondsAgo = Oracle.getMaxSecondsAgo(pool, observationIndex, observationCardinality) / 2;
+        require(metadata.maxSecondsAgo > ORACLE_LOOKBACK, "Aloe: cardinality");
+
+        uint24 fee = pool.fee();
+        metadata.gamma0 = fee;
+        metadata.gamma1 = fee;
+        unchecked {
+            if (feeProtocol % 16 != 0) metadata.gamma0 -= fee / (feeProtocol % 16);
+            if (feeProtocol >> 4 != 0) metadata.gamma1 -= fee / (feeProtocol >> 4);
```

Comment:
I understood most of this function. First, we check the that the data for the provided `metadata` is larger than some time. If it is, we get the instantaneous fee of the `pool` and check if the `feeProtocol` is divisible by 16 and if the `feeProtocol` multiplied by 16 is equal to 0. Why 16?